### PR TITLE
add generational fast path peer metadata APIs

### PR DESCRIPTION
**Pull request by brianolson**
## Description
Add generational API so that when nothing has changed no copy is done (this should be most of the time, just a fast atomic u32 check).
Used in forthcoming network2 refactor.

## Type of Change
- [ ] New feature
- [ ] Bug fix
- [ ] Breaking change
- [X] Performance improvement
- [ ] Refactoring
- [ ] Dependency update
- [ ] Documentation update

## Which Components or Systems Does This Change Impact?
- [X] Validator Node
- [X] Full Node (API, Indexer, etc.)
- [ ] Move/Aptos Virtual Machine
- [ ] Aptos Framework
- [ ] Aptos CLI/SDK
- [ ] Developer Infrastructure
- [ ] Other (specify)

## How Has This Been Tested?
Unit test added. Used in forthcoming network2 branch.

## Key Areas to Review
New unit test does exercise new API.

## Checklist
- [X] I have read and followed the [CONTRIBUTING](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md) doc
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [x] I identified and added all stakeholders and component owners affected by this change as reviewers
- [x] I tested both happy and unhappy path of the functionality
- [x] I have made corresponding changes to the documentation

### CODE REVIEW

File: network/framework/src/application/tests.rs [hunk 1]
```diff
@@ -427,6 +427,102 @@ fn test_peers_and_metadata_caching() {
...
+        Some((pam, newgen)) => {
+            assert_eq!(generation + 1, newgen);
+            assert_eq!(pam.len(), expected_len);
+            newgen
+        },
+        None => {
+            panic!("expected generation 0 -> 1");
+        },
+    }
+}
```
Comment (JoshLind) [inline network/framework/src/application/tests.rs:454]: nit: newline between functions?

File: network/framework/src/application/tests.rs [hunk 2]
```diff
@@ -427,6 +427,102 @@ fn test_peers_and_metadata_caching() {
...
+fn check_peers_and_metadata_generational_none(
+    generation: u32,
+    peers_and_metadata: Arc<PeersAndMetadata>,
+) {
+    if peers_and_metadata
```
Comment (JoshLind) [inline network/framework/src/application/tests.rs:459]: nit: use asserts? e.g.,
`assert!(peers_and_metadata.get_all_peers_generational(...).is_none())`?

File: network/framework/src/application/storage.rs [hunk 3]
```diff
@@ -91,6 +103,69 @@ impl PeersAndMetadata {
         all_peers
     }
 
+    /// Return copy of list of peers if generation is different than previously held.
+    /// May optionally filter for connected peers only.
+    /// May optionally filter on match for any of a set of ProtocolId.
+    pub fn get_all_peers_generational(
```
Comment (JoshLind) [inline network/framework/src/application/storage.rs:109]: Curious: what's the benefit of this over the existing cache? i.e., we currently have a single cache location with cheap reads/writes, but if we move to the generational API, we'll have multiple cached instances (each application will need to maintain a cache of what it last saw and update based on if these APIs return new generations?). Have we seen a bottleneck somewhere to justify moving to the new API?

File: network/framework/src/application/storage.rs [hunk 4]
```diff
@@ -91,6 +103,69 @@ impl PeersAndMetadata {
...
+        generation: u32,
+        require_connected: bool,
+        protocol_ids: &[ProtocolId],
+    ) -> Option<(Vec<PeerNetworkId>, u32)> {
+        let generation_test = self.generation.load(Ordering::SeqCst);
+        if generation == generation_test {
+            return None;
+        }
+        let mut all_peers = Vec::new();
+        let read = self.peers_and_metadata.read();
```
Comment (JoshLind) [inline network/framework/src/application/storage.rs:120]: Silly question: if we move to the generational API, we should remove `cached_peers_and_metadata` entirely, right? Why keep them both around if they're providing the same functionality?

File: network/framework/src/application/storage.rs [hunk 5]
```diff
@@ -91,6 +103,69 @@ impl PeersAndMetadata {
...
+            }
+        }
+        let generation_actual = self.generation.load(Ordering::SeqCst);
+        Some((all_peers, generation_actual))
+    }
+
+    /// Return copy of list of peers if generation is different than previously held.
+    /// May optionally filter for connected peers only.
+    /// May optionally filter on match for any of a set of ProtocolId.
+    pub fn get_all_peers_and_metadata_generational(
```
Comment (JoshLind) [inline network/framework/src/application/storage.rs:140]: Maybe I'm missing the point, but it's tough to see the benefits of this change without actually seeing the call sites change and performance numbers to justify the change. 

Also, if in the next PR we're providing "notifications for connection events", why do we need generational support at all? i.e., applications won't be polling the peers and metadata struct to discover new peers, so reads of peers and metadata information will only be required for "special cases" (e.g., optimizations based on peer metadata), and these reads can already be serviced through the existing cache, no?

Comment (brianolson) [reply network/framework/src/application/storage.rs:140]: The subscription service is just for connects and disconnects; updates to metadata don't generate an event.

**Comment by brianolson** [comment]
'generational' and 'cached' I think were both written to reduce the amount of copying done while holding the lock.
'cached' _always_ does a copy to a public ArcSwap which allows a reader to atomically get _some_ coherent recent value (probably the latest one).
'generational' only does a copy when a reader is using the data and the reader doesn't already have the latest version, and does a copy to the reader's thread local environment that the reader code manages.
Different strategies, probably we only need one. It probably doesn't matter which one. They're probably both good enough and probably both better than the old way and probably a pretty small difference in performance between them. But I wrote mine first and I think it should win ;-)

**Comment by brianolson** [comment]
usage example
https://github.com/brianolson/aptos-core/blob/network2-cd-test/network/framework/src/connectivity_manager/mod.rs#L430-L439

File: network/framework/src/application/storage.rs (hunk 5 above)
Comment (JoshLind) [reply network/framework/src/application/storage.rs:140]: Hmm... that should be fine. Applications don't currently rely on this. Instead, they fetch and read the data when required. It's infrequent for mempool, and more frequent for state sync (but it depends on load). I'm not sure sending events for "metadata changes" provides any benefit.

**Comment by JoshLind** [comment]
Hmm... my thinking is:
1. Given that they both reduce lock contention and (potentially) the number of copies, we shouldn't need/keep both approaches. So, if we want to migrate to the generational APIs, we should completely remove the cache (in the same PR).
2. My concern with the generational API is that we will need to maintain multiple copies of the peers, and multiple copies of the metadata for each application. For example, state sync and [mempool](https://github.com/aptos-labs/aptos-core/pull/12537) rely on the peer and metadata lists. If we move to the generational API, we'll need both applications to maintain an internal cache/copy of the data to use the generational APIs. Consensus will soon be using the metadata, too, so it means we'll end up replacing 1 cache with 3 (and requiring 3 copies of the logic to manage the application "cache"). I think this will become more obvious when we actually update the application code to use the generational API.
3. I'm not that worried about the additional copies of the current approach, e.g., mempool only reads the data every few seconds at startup, and every 10 minutes afterwards, which is pretty rare. State sync reads it more frequently (e.g., specific data requests, but such requests must do a lot more work elsewhere, so the copy is trivial). Without evidence that the current approach is insufficient, I'm not sure the trade-off really makes sense.

Overall, my vote would be to keep the code as-is, and only change it if we see problems arise. @bchocho, @zekun000, any thoughts?

**Comment by JoshLind** [comment]
>usage example
https://github.com/brianolson/aptos-core/blob/network2-cd-test/network/framework/src/connectivity_manager/mod.rs#L430-L439

Is there an example of an application actually using the generational APIs? 😄

**Comment by brianolson** [comment]
This PR is _probably_ going to close. Halfway through reworking network2 branch to use the ArcSwap copy of the data.

**Comment by brianolson** [comment]
generational APIs now gone in nework2. _nevermind_ ...

**Pull request closed without merging**